controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,448 @@
+			"ClientMappingConfig",
+			r.clientMappingConfig.Name,
+		)
+		if err := r.update(r.clientMappingConfig); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update ClientMappingConfig: %v", err)
+		}
+	}
+
+	// Find the StorageClusterPeer from OwnerRef
+	owner := util.FindOwnerRefByKind(r.clientMappingConfig, "StorageClusterPeer")
+	if owner == nil {
+		return ctrl.Result{}, fmt.Errorf("failed to find StorgeClusterPeer owning the ClientMappingConfig")
+	}
+
+	// Fetch the StorageClusterPeer instance
+	r.storageClusterPeer = &ocsv1.StorageClusterPeer{}
+	r.storageClusterPeer.Name = owner.Name
+	r.storageClusterPeer.Namespace = request.Namespace
+
+	if err = r.get(r.storageClusterPeer); err != nil {
+		if k8serrors.IsNotFound(err) {
+			r.log.Info("StorageClusterPeer resource not found. Ignoring since object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get StorageClusterPeer.")
+		return ctrl.Result{}, err
+	}
+
+	if r.storageClusterPeer.Status.State != ocsv1.StorageClusterPeerStatePeered {
+		return ctrl.Result{}, fmt.Errorf(
+			"waiting for StorageClusterPeer %s to be in Peered state",
+			r.storageClusterPeer.Name,
+		)
+	}
+
+	return r.reconcilePhases()
+}
+
+func (r *MirroringReconciler) reconcilePhases() (ctrl.Result, error) {
+	ocsClient, err := providerClient.NewProviderClient(r.ctx, r.storageClusterPeer.Spec.ApiEndpoint, time.Second*10)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create a new provider client: %v", err)
+	}
+	defer ocsClient.Close()
+
+	if err = r.reconcileRBDMirrorDaemon(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if res, err := r.reconcileBlockPoolMirroring(ocsClient); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	if res, err := r.reconcileRadosNamespaceMirroring(ocsClient); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) disableBlockPoolMirroring() (ctrl.Result, error) {
+
+	cephBlockPoolsList, err := r.listCephBlockPools()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for i := range cephBlockPoolsList.Items {
+		cephBlockPool := &cephBlockPoolsList.Items[i]
+
+		cephBlockPool.Spec.Mirroring = rookCephv1.MirroringSpec{}
+		err := r.update(cephBlockPool)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to disable mirroring for CephBlockPool %q: %v", cephBlockPool.Name, err)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) reconcileRBDMirrorDaemon() error {
+	rbdMirror := &rookCephv1.CephRBDMirror{}
+	rbdMirror.Name = util.CephRBDMirrorName
+	rbdMirror.Namespace = r.clientMappingConfig.Namespace
+
+	storageConsumers := &ocsv1alpha1.StorageConsumerList{}
+	if err := r.list(
+		storageConsumers,
+		client.MatchingFields{"annotations": util.RequestMaintenanceModeAnnotation},
+	); err != nil {
+		return fmt.Errorf("failed to list StorageConsumer(s): %v", err)
+	}
+
+	if len(storageConsumers.Items) > 0 {
```

Comment:
This is not a good check, it will always be true (the length will always be greater then 1)